Global tool (jsbp)

tgvashworth opened the pull request:
I've built a little tool that help set up an app by copying a boilerplate into a new directory. I also published it on npm, although there's no docs. (I'll happily remove it if you want to be the publisher)

Try it out:

`npm install -g jsbp`

`jsbp mocha-boilerplate`

:dancers:

Commit:
Demo app not needed.

Commit:
First version of global tool.

Commit:
Bump version up to 0.1.0

Commit:
Merge branch 'feature/global'

tgvashworth commented on bin/jsbp:
```diff
@@ -0,0 +1,184 @@
+    if( !empty ) {
+      program.confirm(msg.not_empty, function (ok) {
+        if( ok ) {
+          fs.remove(path, fn);
```
Looking at it now, I don't thing the target directory should be removed to avoid overwriting existing files.

Commit:
Revert "Demo app not needed."

This reverts commit 8e9751b3afda4df7cdaf08473b65f83f1931e9f9.

rmurphey commented:
FYI, totally not ignoring this, just been holiday-ing over here in the U.S. :) Goal is to have this in some sort of workable state for a talk on Dec 3, so I'll be looking at this tomorrow!

tgvashworth commented:
@rmurphey No problemo, I imagine you've got quite a bit going on right now!

rmurphey commented:
Also, can you add details to the README? No opposition from me that you published the npm module, considering you made it exist :)

tgvashworth commented:
Absolutely, that's my next thang.

Commit:
More feedback to tuser + don't delete their files.

First version deleted the target directory, and didn't tell the user
that it was going to do it. Now it does neither of those.

Commit:
More unified coding style.

Commit:
0.1.1 – Better use feedback & no file deletion!

rmurphey commented:
i am generally +1 to this, except i'm not sure about the behavior when it comes to copying to the destination directory. i haven't really been sure how to handle this -- should it create a `mocha-boilerplate` dir in the destination dir? i don't think it's right for it to just dump all of the subdirs into the dir, as it's doing now, but I'm not sure what the right answer is, either. perhaps the boilerplate itself should be restructured so it _is_ safer to dump everything into the destination dir? interested in your thoughts, @phuu --

tgvashworth commented:
I was going to ask you the same question last night :) Should it create/copy into a directory with the name of the app, but put the boilerplate into a subdir, with a boilerplate 'app' folder alongside?

rmurphey commented:
A few assumptions I'm making here:
- the user should not be forced to use requirejs for their app, but will be required to use it for their tests
- the user should not be forced to use grunt for their app, and if they are using it, the boilerplate grunt file should not overwrite their existing grunt file
- it should be possible to update the boilerplate by updating the lib and runner dirs

I think in the case of creating a new app `myApp`:
- create the `myApp` dir
- create a dir `myApp/mocha-boilerplate` (lib, tests, runner, etc)

In the case of working with an existing app dir `myApp`:
- create a dir `myApp/mocha-boilerplate` with the boilerplate's contents (lib, tests, runner, etc)

However, I hate the idea of having a dir named `mocha-boilerplate` inside my project. I want to call it `tests`, but then you have `tests/tests` and that sucks. Thoughts?

tgvashworth commented:
So here's my thinking: in the case of a Mocha boilerplate it does make sense to copy into a subfolder, but what if there were a boilerplate that needed app files set out in a certain way?

I think each boilerplate should be it's own isolated project, so that if a different file layout made sense then you could just create the boilerplate for that, and add it to the list.

For example:

```
stupid-example-bp/
    README.md
    index.html
    app/
       app.js
    test/
       test.js

another-example-bp/
    README.md
    js/
        app.js
    vendor/
        jquery.js
    tests/
        test.js
```

That way, jsbp can just duplicate the directory and it works straight away, plus the instructions can be per-boilerplate, and adding a new one is a simple as adding it to the list. IMO that's the simplest option, with the least surprises. What do you think?

tgvashworth commented:
Just to say I haven't forgotten about this, just haven't had time to work on it. Gonna try and fit some time in next week.

tgvashworth commented:
@rmurphey Do you have an opinion on this?

rmurphey commented:
I've had a lot of conversations with a lot of people in the last couple weeks surrounding testing ... I'm not sure what the best way forward is. 

What I _want_ to accomplish with this is a single dir (per testing framework) that can be dropped into an _existing_ project in order to allow developers to start writing tests. This dir should include lib files (mocha, sinon, etc), an html runner, and a dir for tests. It should be structured in such a way that we can "upgrade" the libs and the runner without the user having to modify their tests (assuming there are no API changes in the libs). 

All of that is to say, I want to be pretty agnostic about what _else_ someone has going on in their project -- this should be a single dir that can be added to a project to allow for testing, it should be self-contained, and the boilerplate parts should be updateable.

The examples above seem to make assumptions about the app; I really want to stay away from that if possible. 

For an example of this essentially working in an existing project (though I ended up using a shared config between the tests and the app), see [this demo](https://github.com/rmurphey/testable-javascript).
